dcos/package.py

```diff
@@ -19,40 +19,43 @@ def uninstall(pkg, package_name, remove_all, app_id, cli, app):
     :type remove_all: boolean
     :param app_id: App ID of the app instance to uninstall
     :type app_id: str
-    :param init_client: The program to use to run the app
-    :type init_client: object
+    :param cli: Whether to remove the CLI, setting it to None will let
+                the program automatically decide (depending on the number
+                of app instances to remain after uninstallation)
+    :type remove_all: boolean
+    :param app: Whether to remove app
+    :type remove_all: boolean
     :rtype: None
     """
 
-    if cli is False and app is False:
-        cli = app = True
-
-    uninstalled = False
     installed = installed_packages(
-        pkg, app_id, package_name, cli_only=False)
-    installed_cli = next((True for installed_pkg in installed
-                          if installed_pkg.get("command")), False)
-    installed_app = next((True for installed_pkg in installed
-                          if installed_pkg.get("apps")), False)
-
-    if cli and installed_cli:
-        if subcommand.uninstall(package_name):
-            uninstalled = True
-
-    if app and installed_app:
-        if pkg.uninstall_app(package_name, remove_all, app_id):
-            uninstalled = True
+        pkg, None, package_name, cli_only=False)
+    installed_pkg = next(iter(installed), None)
+    installed_cli = installed_pkg.get("command") if installed_pkg else None
+    installed_app = installed_pkg.get("apps") if installed_pkg else []
 
-    if uninstalled:
-        return None
-    else:
+    if not installed_app and not installed_cli:
         msg = 'Package [{}]'.format(package_name)
         if app_id is not None:
             app_id = util.normalize_marathon_id_path(app_id)
             msg += " with id [{}]".format(app_id)
         msg += " is not installed"
         raise DCOSException(msg)
 
+    # If cli is set to None, try to be conservative and
+    # delete the CLI only when no apps would remain.
+    if cli is None:
+        cli = (remove_all or not app_id
+                or installed_app.count(app_id) == len(installed_app))
```

Comment:
I find this logic hard to follow (see comment below for a suggested alternative).